### ISSUE TITLE
Can't restore due to missing archive, but archive didn't exist in first place

### ISSUE THREAD

**Issue by WillPlatnick** [wal-g/wal-g#121]
We are using WAL-G on 4 different database clusters. We're having issues restoring the database from this one cluster after an upgrade to PG10.  This archive that it's looking for doesn't exist in S3.

The WAL log it's looking for is ```000000010000D6E5000000C1``` and it's true, it doesn't exist.

In fact, that whole series doesn't exist. I have no files that start with ```000000010000D6``` but I do have a ton with ```000000010000D7```

I'm wondering if something with the timeline is getting messed up during backups or during restore somehow, but I'm not sure where to look first.

```2018-09-22 13:57:51.549 GMT [2278] LOG:  creating missing WAL directory "pg_wal/archive_status"
2018-09-22 13:57:51.587 GMT [2278] LOG:  starting point-in-time recovery to 2018-09-22 07:00:00+00
BUCKET: bucketname
SERVER: 
2018/09/22 08:57:52 Archive '%s' does not exist 000000010000D6E5000000C1
github.com/wal-g/wal-g.DownloadAndDecompressWALFile
	/home/travis/gopath/src/github.com/wal-g/wal-g/commands.go:590
github.com/wal-g/wal-g.HandleWALFetch
	/home/travis/gopath/src/github.com/wal-g/wal-g/commands.go:515
main.main
	/home/travis/gopath/src/github.com/wal-g/wal-g/cmd/wal-g/main.go:131
runtime.main
	/home/travis/.gimme/versions/go1.10.4.linux.amd64/src/runtime/proc.go:198
runtime.goexit
	/home/travis/.gimme/versions/go1.10.4.linux.amd64/src/runtime/asm_amd64.s:2361
2018-09-22 13:57:52.014 GMT [2278] LOG:  invalid checkpoint record
2018-09-22 13:57:52.014 GMT [2278] FATAL:  could not locate required checkpoint record
2018-09-22 13:57:52.014 GMT [2278] HINT:  If you are not restoring from a backup, try removing the file "/var/lib/postgresql/10/main/backup_label".```

**Comment by WillPlatnick** [comment]
To add more context, it's been the same issue for the past 4 days, full backups daily. Each time, it's the same thing, it's looking for a log file whose entire series doesn't exist, and the real series is always incremented by one.

IE, looking for ```000000010000D7450000005C``` in restore

```000000010000D74``` isn't a series but ```000000010000D75``` is.

There's nothing in my restore logs or my postgresql logs with the word ```timeline```

**Comment by x4m** [comment]
Hi! 
Plz check that there is no backup_label file in master. Currently, I have no other explanations...
Timeline here is 00000001, D7455C is segment... btw, what's the name of the backup which tries to load 000000010000D7450000005C?

**Comment by x4m** [comment]
Well, if backup is named ```D74500000059``` you need all WAL files since ```D74500000059``` for PITR and at least some of them for plain restoration.
I know the only legitimate case when WALs can be not archived by WAL-G --  pg_upgrade. In this case backup that you are restoring is from version before 10. You can check file PG_VERSION in PGDATA to test this version.

if it's not a case - we need to figure out how come that WALs are in in archive.

**Comment by WillPlatnick** [comment]
We used pg_upgrade on Friday. We've run 4 backup and restore operations since then. There are no wal logs in series ```D74``` only ```D75````. PG_VERSION is 10 as it should be. The log that it's looking for during the restore is the very first log.

**Comment by x4m** [comment]
Did you change WAL-G version or something in environment?

**Comment by x4m** [comment]
Did you use ```wal-g delete`` command to clean up archives?

**Comment by WillPlatnick** [comment]
Nope, upgrades were upgraded in same way across all clusters, WAL-G version stayed the same. We run ```delete retain FULL 7 --confirm``` each day for every cluster.

**Comment by x4m** [comment]
Do I understand correctly following?
1. You observe the problem on many clusters
2. You are restoring LATEST backup?
3. The problem reproduce with new backups each day

Can you check existance of WAL 00000001000XXX... right after backup base_00000001000XXX... will be created? And check it once more right after delete. And...well...check again before backup restoration.

**Comment by WillPlatnick** [comment]
I'm taking a backup now. It didn't increment anything crazily unexpected. In my postgres file system, I see a .done in archive_status for the log it's looking for, so, my only guess is the delete operation is deleting things it shouldn't. I've turned off the deletion and we'll see how it goes.

Now why it would be doing it only for this job, when other restore jobs happen later in the day and are fine after the same delete operation. This is the write-heaviest of the clusters, maybe a wrap-around issue? We'll see tomorrow.

**Comment by x4m** [comment]
whoa... that's scary, will look into this too

**Comment by x4m** [comment]
WAL-G must hold this invariant: we delete WAL only when it's less than every existent backup. If it messes even LATEST backup - that's not good. But I do not see any reproduction on my systems. Neither I see bugs in the code...

**Comment by x4m** [comment]
@WillPlatnick can I ask you to list which files were in place before delete and which left after delete?
Also, can you attach the output of delete command?

**Comment by WillPlatnick** [comment]
Sorry for the late response, I had another issue with our restore that popped up with PG10 that I had to get around before I could run delete operations again.

First, a list of backups

```
SERVER:
name                          last_modified        wal_segment_backup_start
base_000000030000D681000000A4 2018-09-19T07:02:33Z 000000030000D681000000A4
base_000000030000D6A0000000F8 2018-09-20T07:04:39Z 000000030000D6A0000000F8
base_000000040000D6C400000053 2018-09-21T06:56:27Z 000000040000D6C400000053
base_000000010000D6E5000000BF 2018-09-22T06:57:58Z 000000010000D6E5000000BF
base_000000010000D7010000003E 2018-09-23T06:55:36Z 000000010000D7010000003E
base_000000010000D727000000F0 2018-09-24T06:56:56Z 000000010000D727000000F0
base_000000010000D74500000059 2018-09-25T07:01:28Z 000000010000D74500000059
base_000000010000D76200000099 2018-09-26T06:56:44Z 000000010000D76200000099
base_000000010000D785000000BA 2018-09-27T06:57:46Z 000000010000D785000000BA
base_000000010000D7A500000063 2018-09-28T06:56:57Z 000000010000D7A500000063
base_000000010000D7C100000060 2018-09-29T06:57:16Z 000000010000D7C100000060
base_000000010000D7E400000081 2018-09-30T06:57:23Z 000000010000D7E400000081
```

Notice the shifted timelines? I wonder if the bug had something to do with that. I just ran the delete operation again, and I see the files I expect in the object store. It seems my timelines are now at 1 since pgupgrade. This delete got rid of the last full's that had different timelines, so I'm going to turn my deletes on again and see how we do.

**Comment by x4m** [comment]
@WillPlatnick now it's clear that it's a bug. We compare WALs by timelines, but timelines can be resetted (during pg_upgrade). When WAL-G picks 7th backup to keep, it takes backup from previous pg version. And it has bign TL number. So all current WALs are considered as long before those 7th backup...

I'll address it in nearest days.
I wonder what would be correct deletion policy for WALs... and base backups too.

**Comment by x4m** [comment]
We did not encounter this situation in our testing because we switch S3 bucket when upgrade...

**Comment by dev1ant** [comment]
Shouldn't we mention in the documentation that different major versions should have different paths in S3?

Of course that doesn't mean that we shouldn't make some kind of protection from this configuration.

**Comment by x4m** [comment]
there are few ways to implement this warning:
1. Add bucket naming considerations to README.md. PR is welcome, especially from native English speakers.
2. Support config option of PG_VERSION that will store WALs from different pg versions separately. This will complicate deletion logic, but will be much safer in this aspect.

**Comment by x4m** [comment]
Will, I've pushed proposed fix: WAL-G will calculate minimum WalFileName (TL+LSN) among all skipped backups and use it instead of WalFileName of the last backup. This, effectively, will enforce skipping all WALs in case of upgrade until you get away on 7 backups from upgrade.
If you have some spare time plz review commit, if everything is OK I'll backport this to REL0_1_STABLE.
Thanks!
